engine/consensus/matching/engine.go

```diff
@@ -98,1063 +75,169 @@ func New(
-	added, err := e.incorporatedResults.Add(
-		flow.NewIncorporatedResult(
-			receipt.ExecutionResult.BlockID,
-			&receipt.ExecutionResult,
-		),
-	)
-	if err != nil {
-		return fmt.Errorf("error inserting incorporated result in mempool: %w", err)
-	}
-	if !added {
-		log.Debug().Msg("skipping result already in mempool")
-		return engine.NewDuplicatedEntryErrorf("")
-	}
-	e.mempool.MempoolEntries(metrics.ResourceResult, e.incorporatedResults.Size())
-	return nil
-}
-
-// onApproval processes a new result approval.
-func (e *Engine) onApproval(originID flow.Identifier, approval *flow.ResultApproval) error {
-	startTime := time.Now()
-	approvalSpan := e.tracer.StartSpan(approval.ID(), trace.CONMatchOnApproval)
-	defer func() {
-		e.metrics.OnApprovalProcessingDuration(time.Since(startTime))
-		approvalSpan.Finish()
-	}()
-
-	log := e.log.With().
-		Hex("approval_id", logging.Entity(approval)).
-		Hex("block_id", approval.Body.BlockID[:]).
-		Hex("result_id", approval.Body.ExecutionResultID[:]).
-		Logger()
-	log.Info().Msg("result approval received")
-
-	// Check that the message's origin (as established by the networking layer) is
-	// equal to the message's creator as reported by the message itself. Thereby,
-	// we rely on the networking layer for enforcing message integrity via the
-	// networking key.
-	if approval.Body.ApproverID != originID {
-		return engine.NewInvalidInputErrorf("invalid origin for approval: %x", originID)
-	}
-
-	// check if we already have the block the approval pertains to
-	head, err := e.state.AtBlockID(approval.Body.BlockID).Head()
-	if err != nil {
-		if !errors.Is(err, storage.ErrNotFound) {
-			return fmt.Errorf("failed to retrieve header for block %x: %w", approval.Body.BlockID, err)
-		}
-		// Don't error if the block is not known yet, because the checks in the
-		// else-branch below are called again when we try to match approvals to chunks.
-	} else {
-		// drop approval, if it is for block whose height is lower or equal to already sealed height
-		sealed, err := e.state.Sealed().Head()
-		if err != nil {
-			return fmt.Errorf("could not find sealed block: %w", err)
-		}
-		if sealed.Height >= head.Height {
-			log.Debug().Msg("discarding approval for already sealed and finalized block height")
-			return nil
-		}
-
-		// Check if the approver was a staked verifier at that block.
-		err = e.ensureStakedNodeWithRole(approval.Body.ApproverID, head, flow.RoleVerification)
-		if err != nil {
-			return fmt.Errorf("failed to process approval: %w", err)
+// processEvents is processor of pending events which drives events from networking layer to business logic in `Core`.
+// Effectively consumes messages from networking layer and dispatches them into corresponding sinks which are connected with `Core`.
+// Should be run as a separate goroutine.
+func (c *Engine) processEvents() {
+	fetchEvent := func(queue *deque.Deque, sink EventSink) (*Event, EventSink) {
```

Comment:
why do you need the sink here?

Reply:
Here we select item and sink, if there is no item then we shouldn't add it as part of select statement. That's just an easy way to get (item, sink) or (nil and nil).